### PR TITLE
Initial human-readable logging

### PR DESCRIPTION
**Pull request by breerly**
This PR makes Fx output a series of uniform and human readable output following these principles:

* Should only rely on stdlib
* Should be line-oriented (no console tables)
* Should be greppable
* Should only output what is necessary

The following output answers these 3 basic questions:

* "What types are available in the container?" - scan the PROVIDE lines
* "What types were invoked and in what order?" scan the INVOKE lines
* "What lifecycle events executed, and in what order?" scan the START/STOP lines

```
devexp/fx-playground.git - [yarpc-revamp●] » ./turnkey 2>&1 | grep 'Fx'
2017/05/26 11:51:47 [Fx] PROVIDE        *fx.Lifecycle
2017/05/26 11:51:47 [Fx] PROVIDE        *servicefx.ServiceDetail <= lyft.com/example.git/external/modules/servicefx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        *zap.Logger <= lyft.com/example.git/external/modules/zapfx.(*Module).Module-fm()
2017/05/26 11:51:47 [Fx] PROVIDE        config.Provider <= lyft.com/example.git/external/modules/configfx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        tally.Scope <= lyft.com/example.git/external/modules/metricsfx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        tally.CachedStatsReporter <= lyft.com/example.git/external/modules/metricsfx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        opentracing.Tracer <= lyft.com/example.git/external/modules/tracerfx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        *rpcinit.Dispatcher <= lyft.com/example.git/external/modules/yarpcfx.Module()
2017/05/26 11:51:47 [Fx] PROVIDE        *[]transport.Procedure <= lyft.com/example.git/handler.NewProcedures()
2017/05/26 11:51:47 [Fx] PROVIDE        turnkey.Client <= lyft.com/example.git/client/turnkey.New()
2017/05/26 11:51:47 [Fx] PROVIDE        *handler.Handler <= lyft.com/example.git/handler.New()
2017/05/26 11:51:47 [Fx] PROVIDE        *handler.HelloHandler <= lyft.com/example.git/handler.NewHelloHandler()
2017/05/26 11:51:47 [Fx] PROVIDE        *handler.HiHandler <= lyft.com/example.git/handler.NewHiHandler()
2017/05/26 11:51:47 [Fx] PROVIDE        *handler.FarewellHandler <= lyft.com/example.git/handler.NewFarewellHandler()
2017/05/26 11:51:47 [Fx] PROVIDE        *yarpc.InboundMiddleware <= lyft.com/example.git/middleware.NewInboundMiddleware()
2017/05/26 11:51:47 [Fx] PROVIDE        *middleware.LoggingUnaryInbound <= lyft.com/example.git/middleware.NewLoggingUnaryInbound()
2017/05/26 11:51:47 [Fx] INVOKE         lyft.com/example.git/external/modules/yarpcfx.Run()
2017/05/26 11:51:47 [Fx] INVOKE         main.newBiz()
2017/05/26 11:51:47 [Fx] START          lyft.com/example.git/external/modules/yarpcfx.Module()
2017/05/26 11:51:47 [Fx] ERROR          Start failed, rolling back: listen tcp :8081: bind: address already in use
2017/05/26 11:51:47 [Fx] STOP           lyft.com/example.git/external/modules/tracerfx.Module()
2017/05/26 11:51:47 [Fx] STOP           lyft.com/example.git/external/modules/zapfx.(*Module).Module()
2017/05/26 11:51:47 [Fx] STOP           lyft.com/example.git/external/modules/configfx.Module()
2017/05/26 11:51:47 [Fx] ERROR          Couldn't rollback cleanly: sync /dev/stderr: operation not supported
2017/05/26 11:51:47 [Fx] ERROR          Failed to start: listen tcp :8081: bind: address already in use; sync /dev/stderr: operation not supported
```

### CODE REVIEW

**Comment by breerly** [comment]
This is a big improvement in user experience, without adding too much weight to Fx's core types.

I'd like to add this now, without too much scrutiny, and then iterate as we hone in on 1.0.

File: internal/fxreflect/fxreflect.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+// FuncName returns a funcs formatted name
+func FuncName(fn interface{}) string {
+	fnName := runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
+	return fmt.Sprintf("%s()", fnName)
+}
+
+func isErr(t reflect.Type) bool {
+	errInterface := reflect.TypeOf((*error)(nil)).Elem()
+	if t.Implements(errInterface) {
```
Comment (alsamylkin) [inline internal/fxreflect/fxreflect.go]: just `return t.Implements(errInterface)`?

File: internal/fxlog/fxlog.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fxlog
```
Comment (glibsm) [inline internal/fxlog/fxlog.go]: no scrutiny, just a question: why package-level functions and not a logging object embedded in the App?

File: internal/fxreflect/fxreflect.go (hunk 1 above)
Comment (breerly) [inline internal/fxreflect/fxreflect.go]: doh

File: lifecycle.go [hunk 3]
```diff
@@ -43,6 +50,7 @@ type lifecycle struct {
 }
 
 func (l *lifecycle) Append(hook Hook) {
+	hook.caller = fxreflect.Caller()
```
Comment (glibsm) [inline lifecycle.go]: this is great

File: internal/fxlog/fxlog.go (hunk 2 above)
Comment (breerly) [inline internal/fxlog/fxlog.go]: we'll def want that - just didn't want to overhaul it in this PR

File: lifecycle.go (hunk 3 above)
Comment (breerly) [inline lifecycle.go]: was a PITA :P

File: internal/fxlog/fxlog.go (hunk 2 above)
Comment (breerly) [inline internal/fxlog/fxlog.go]: thats actually needed to do a thorough test of the whole deal... will submit another PR that makes that change.
